torch/fx/experimental/symbolic_shapes.py

```diff
@@ -2827,11 +2876,30 @@ def get_symbol(tensor_dim_src):
                 concrete_val = self.evaluate_expr(sympy.Eq(s1, s2))
                 if not concrete_val:
                     raise ConstraintViolationError(
-                        f"{src1.name()} = {self.var_to_val[s1]}"
+                        f"{src1.name()} = {s1.subs(self.var_to_val)}"
                         " is not equal to "
-                        f"{src2.name()} = {self.var_to_val[s2]}"
+                        f"{src2.name()} = {s2.subs(self.var_to_val)}"
                     )
 
+            for src, root, fn in equalities_inputs.derived_equalities:
+                s1 = get_symbol(src)
+                s2 = root if isinstance(root, sympy.Symbol) else get_symbol(root)
+                s2_ = fn(s2)
+                concrete_val = self.evaluate_expr(sympy.Eq(s1, s2_))
```

Comment:
evaluate_expr feels like the wrong thing to use here? Because it conventionally will install a guard if we don't know if something is statically the case. Maybe you want `maybe_static_evaluate` instead?

Reply:
OK, you might be right: if I really want to prove something given the guards, do static evaluation.

Reply:
So I remembered what this is about, and we do need the `evaluate_expr` here: we want to add a guard if necessary. Specifically, this is checking whether concrete inputs satisfy the specified condition. We're not proving anything.